Autodesk: General HgiVulkan stability

erikaharrison-adsk opened the pull request:
### Description of Change(s)

This PR provides the following support to the hgiVulkan backend
1. Fixes Color Correction
2. Provides Layout Transition infrastructure
3. Fixes most Vulkan validation errors
4. Fixes debug layer struct issue
5. Fixes Pipeline Vertex Divisor struct issue

### Fixes Issue(s)
- N/A

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

VipulKapoorAutodesk committed:
This commit provides the following to hgiVulkan path
1. Fixes Color Correction
2. Provides Layout Transition infrastructure
3. Fixes most Vulkan validation errors
4. Fixes debug layer struct issue

(cherry picked from commit d10ea20a202458727ae5d667dae5dd9bf9bdfbf7)

schevrel commented on pxr/imaging/hdx/colorCorrectionTask.cpp:
```diff
@@ -877,11 +877,19 @@ HdxColorCorrectionTask::_CreateBufferResources()
               { -1, -1, 0, 1,     0, 0 },
               {  3, -1, 0, 1,     2, 0 } };
 
+    // Vulkan back-end needs the UVs inverted along the y axis
```
Instead of fixing this here, it might be better to bring in the fix at the shader generation level.  See https://github.com/PixarAnimationStudios/OpenUSD/pull/2504

schevrel replied:
My PR had some issues with HdStorm, I have updated it to remove some shader compilation errors.

jesschimein commented:
Filed as internal issue #USD-9060

clach commented on pxr/imaging/hgiVulkan/texture.cpp:
```diff
@@ -514,6 +522,73 @@ HgiVulkanTexture::CopyBufferToTexture(
+    // The following cases are based on few initial assumptions to provide
+    // an infrastructure for access mask selection based on layouts.
+    // Feel free to update depending on need and use cases.
+    switch (GetImageLayout()) {
```
This and the following switch statement need a default case, otherwise they fail to compile for me.

VipulKapoorAutodesk replied:
Fixed.

clach commented on pxr/imaging/hgiVulkan/graphicsPipeline.cpp:
```diff
@@ -132,21 +132,19 @@ HgiVulkanGraphicsPipeline::HgiVulkanGraphicsPipeline(
         vertBufs.push_back(std::move(vib));
     }
 
+    VkPipelineVertexInputDivisorStateCreateInfoEXT vertexInputDivisor =
```
I'm guessing this work was done on an earlier version of the dev/feature branch, since I believe this change is no longer needed.

VipulKapoorAutodesk replied:
Sure. This will be undone in the oncoming push

clach commented on pxr/imaging/hgiVulkan/instance.cpp:
```diff
@@ -67,15 +67,15 @@ HgiVulkanInstance::HgiVulkanInstance()
         VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME,
     };
 
+    std::vector<const char*> instanceLayerList;
```
This I think is also unneeded now.

clach commented on pxr/imaging/hdSt/interleavedMemoryManager.cpp:
```diff
@@ -462,8 +462,11 @@ HdStInterleavedMemoryManager::_StripedInterleavedBuffer::Reallocate(
-        bufDesc.usage = HgiBufferUsageUniform;
+        // Vulkan Validation layer is raising an error here because
+        // a usage flag of Uniform is applied to a buffer bound as storage
+        bufDesc.usage = HgiBufferUsageUniform | HgiBufferUsageStorage;
```
If you want this to run with HgiGL, too, then can you add a Vulkan-specific check here like you did for vboMemoryManger.cpp?

clach commented on pxr/imaging/hdSt/vboSimpleMemoryManager.cpp:
```diff
@@ -292,7 +292,9 @@ HdStVBOSimpleMemoryManager::_SimpleBufferArray::Reallocate(
-        bufDesc.usage = HgiBufferUsageUniform | HgiBufferUsageVertex;
+        // Vulkan Validation layer is raising an error here because
+        // a usage flag of Uniform is applied to a buffer bound as storage
+        bufDesc.usage = HgiBufferUsageUniform | HgiBufferUsageStorage;
```
Same here re: a Vulkan-specific check so this works for HgiGL too.

VipulKapoorAutodesk committed:
some code clean up.

(cherry picked from commit 173d02c573ed820e4d79aa3aebe5610298963d04)

clach commented on pxr/imaging/hdSt/codeGen.cpp:
```diff
@@ -147,8 +147,14 @@ TF_DEFINE_PRIVATE_TOKENS(
     (early_fragment_tests)
 );
 
+// For now, looks like for Vulkan path, we require HGI Resource Generation path enabled
```
This too is already fixed in the current branch, just FYI.

VipulKapoorAutodesk committed:
cleaned up code around
1. VkPipelineVertexInputStateCreateInfo (following latest fixes proposed by Pixar)
2. Instance Layer Extensions (following latest fixes proposed by Pixar)
3. UV flipping fix in shader selection for Vulkan (following latest fixes proposed by Adobe)
4. Undoing temp code aound flipping color correction UVs flip

(cherry picked from commit db0a46bede5b9504bb6954dc8ac7b1b5d6e5f364)

clach commented on pxr/imaging/hgiVulkan/device.cpp:
```diff
@@ -253,7 +254,7 @@ HgiVulkanDevice::HgiVulkanDevice(HgiVulkanInstance* instance)
     VkPhysicalDeviceFeatures2 features =
         {VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FEATURES_2};
-    features.pNext = &vulkan11Features;
+    features.pNext = (instance->GetVulkanVersion() >= VK_API_VERSION_1_2) ? &vulkan11Features : nullptr;
```
Can you elaborate on the need for this change? I'm assuming this is because you're potentially targeting Vulkan versions < 1.2 for Android support. Do you have an idea of what other changes will be needed for that goal?

Right now we use vulkan11Features here to potentially enable shaderDrawParameters (you can see the associated hgi capability flag set in HgiVulkanCapabilities). If we potentially disapply vulkan11Features here, would that need a corresponding modification to HgiVulkanCapabilities?
